tests/structured_conf/test_structured_config.py

```diff
@@ -1144,3 +1144,191 @@ def test_merge_missing_list_promotes_target_type(self, class_type: str) -> None:
+        self, class_type: str, obj: str, value: Any, node: str
+    ) -> None:
+        module: Any = import_module(class_type)
+        class_ = getattr(module, obj)
+        cfg = OmegaConf.structured(class_)
+        cfg[node] = value
+        assert cfg[node] == value
+
+    @pytest.mark.parametrize(
+        "obj,value,node",
+        [
+            ("ContainerInList", [{"user": True}], "list_with_dict"),
+            ("ContainerInList", [{"user": User()}], "list_with_dict"),
+            ("ContainerInList", [[1, 2], [True, 4]], "list_with_list"),
+            ("ContainerInList", [[1, 2], [User(), 4]], "list_with_list"),
+            ("ContainerInDict", {"users": {"default": True}}, "dict_with_dict"),
+            ("ContainerInDict", {"users": {"default": User()}}, "dict_with_dict"),
+            ("ContainerInDict", {"users": [1, True]}, "dict_with_list"),
+            ("ContainerInDict", {"users": [1, User()]}, "dict_with_list"),
+            ("ComplexList", [[{"number": True}]], "list"),
+            ("ComplexList", [[{"number": User()}]], "list"),
+            ("ComplexDict", {"users": {"data": True}}, "dict"),
+            ("ComplexDict", {"users": {"data": User()}}, "dict"),
+        ],
+    )
+    def test_container_as_element_type_invalid(
+        self, class_type: str, obj: str, value: Any, node: str
+    ) -> None:
+        module: Any = import_module(class_type)
+        class_ = getattr(module, obj)
+        cfg = OmegaConf.structured(class_)
+        with pytest.raises(ValidationError):
+            cfg[node] = value
+
+    @pytest.mark.parametrize(
+        "obj,obj2,expected",
+        [
+            (
+                "ContainerInList",
+                {"list_with_list": [[4, 5]], "list_with_dict": [{"foo": 4}]},
+                {"list_with_list": [[4, 5]], "list_with_dict": [{"foo": 4}]},
+            ),
+            (
+                "ContainerInDict",
+                {
+                    "dict_with_list": {"foo": [3, 4]},
+                    "dict_with_dict": {"foo": {"var": 4}},
+                },
+                {
+                    "dict_with_list": {"foo": [3, 4]},
+                    "dict_with_dict": {"foo": {"var": 4}},
+                },
+            ),
+            (
+                "ComplexList",
+                {"list": [[{"foo": 13}]]},
+                {"list": [[{"foo": 13}]]},
+            ),
+            (
+                "ComplexDict",
+                {"dict": {"foo2": {"var2": [3, 4]}}},
+                {"dict": {"foo": {"var": [1, 2]}, "foo2": {"var2": [3, 4]}}},
+            ),
+        ],
+    )
+    def test_container_as_element_type_merge(
+        self, class_type: str, obj: str, obj2: Any, expected: Any
+    ) -> None:
+        module: Any = import_module(class_type)
+        class_ = getattr(module, obj)
+        cfg = OmegaConf.merge(class_)
+        res = OmegaConf.merge(cfg, obj2)
+        assert res == expected
+
+    @pytest.mark.parametrize(
+        "obj,obj2",
```

Comment:
the first field here is a class name (class_type), not an object. looks like this mistake is repeated in all the new tests here.